### PR TITLE
Pnv contributions

### PR DESCRIPTION
**Pull request by chibenwa**
https://www.passerellesnumeriques.org/en/

The results of Wednesday practical work (8h).

I will create Apache tickets later on.

With many thanks to: 
 - Nguyen Thi Mai and Le Thi Huong Lai for their work on Quotas and In memory subscriptions
 - Tran Thi & My Linh for their work on TooMuchLines
 - txc1996 &  Van Thanh For improving readability of SMTPAuthIsSuccessfulTest
 - thienan090196 &  Tran Thi & My Linh for their tests on RecipientIsLocalTest

Many other students in PNV  contributed to this doing MOOB programming.

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+
+import javax.mail.MessagingException;
+
+import org.apache.mailet.Mail;
+import org.apache.mailet.MailAddress;
+import org.apache.mailet.base.GenericMatcher;
+
+public class TooMuchLines extends GenericMatcher {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java]: you have to use *many* when the thing is countable.

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+		maximumLineCount = Integer.valueOf(condition);
+		if (maximumLineCount < 1) {
+		    throw new MessagingException("Condition should be strictly positive");
+		}
+	}
+
+	@Override
+	public Collection<MailAddress> match(Mail mail) throws MessagingException {
+		if(mail.getMessage() == null){
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java]: missing spaces

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		    return new ArrayList<MailAddress>();
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java]: could you return an ImmutableList instead ?

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    public void setUp() {
+        testee = new TooMuchLines();
+    }
+
+    @Test(expected = MessagingException.class)
+    public void initShouldThrowOnAbsentCondition() throws Exception {
+        testee.init(FakeMatcherConfig.builder().matcherName("name").build());
+    }
+
+    @Test(expected = NumberFormatException.class)
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java]: it should throw a more james specific exception

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            .mimeMessage(MimeMessageBuilder.mimeMessageBuilder()
+                .setMultipartWithBodyParts(MimeMessageBuilder.bodyPartBuilder()
+                    .data("content")
+                    .build())
+                .build())
+            .build();
+
+        Collection<MailAddress> result = testee.match(fakeMail);
+
+        assertThat(result).isEmpty();
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java]: aren't we supposed the retrieve MailAdresses here ?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import org.apache.james.mailbox.store.user.model.Subscription;
+import org.apache.james.mailbox.store.user.model.impl.SimpleSubscription;
+import org.junit.Before;
+import org.junit.Test;
+
+public class InMemorySubscriptionMapperTest {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java]: what is specific to InMemory ?

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SMTPAuthSuccessfulTest.java [hunk 7]
```diff
@@ -17,29 +17,53 @@
...
+        testee.init(FakeMatcherConfig.builder().matcherName("matcherName")
+            .mailetContext(FakeMailContext.defaultContext())
+            .build());
     }
 
-    protected String getMatcherName() {
-        return "SMTPAuthSuccessful";
+    @Test
+    public void matchShouldReturnRecipientsWhenAuthUserAttributeIsPresent() throws Exception{
+        MailAddress recipient = new MailAddress("cuong.trn@gmail.com");
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SMTPAuthSuccessfulTest.java]: is it a real address ?

**Comment by chibenwa** [comment]
test this please

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java (hunk 3 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java]: I catched a lot but I missed this one

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java]: Nothing. I will extract a more generic class of test from their work.

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SMTPAuthSuccessfulTest.java (hunk 7 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SMTPAuthSuccessfulTest.java]: Looks like. I modified some of the address they wrote by I guess I missed this one.

**Comment by chibenwa** [comment]
Thanks @mbaechler for the review

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java (hunk 5 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/TooMuchLinesTest.java]: The mail don't have too much lines, then we should not match ;-). So we return empty

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    public void findSubscriptionsForUserShouldReturnOnlyUserSubscriptions() {
+        SimpleSubscription subscription1 = new SimpleSubscription(USER_1,MAILBOX_1);
+        SimpleSubscription subscription2 = new SimpleSubscription(USER_2,MAILBOX_2);
+        testee.save(subscription1);
+        testee.save(subscription2);
+
+        List<Subscription> results = testee.findSubscriptionsForUser(USER_1);
+
+        assertThat(results).containsOnly(subscription1);
+
```
Comment (quynhn) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java]: too many while lines here

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsOverQuota.java [hunk 9]
```diff
@@ -23,41 +44,39 @@
...
             for (MailAddress mailAddress : mail.getRecipients()) {
-                MailboxSession mailboxSession = mailboxManager.createSystemSession(mailAddress.getLocalPart(), LOGGER);
+                String userName = usersRepository.getUser(mailAddress);
+                MailboxSession mailboxSession = mailboxManager.createSystemSession(userName, LOGGER);
                 MailboxPath mailboxPath = MailboxPath.inbox(mailboxSession);
                 QuotaRoot quotaRoot = quotaRootResolver.getQuotaRoot(mailboxPath);
-                if (quotaManager.getMessageQuota(quotaRoot).isOverQuota() &&
-                    quotaManager.getStorageQuota(quotaRoot).isOverQuota()) {
+
+                if (quotaManager.getMessageQuota(quotaRoot).isOverQuotaWithAdditionalValue(1) ||
```
Comment (quynhn) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsOverQuota.java]: extract "1" to constant and naming it

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/TooManyLinesTest.java [hunk 10]
```diff
@@ -55,7 +55,7 @@ public void initShouldThrowOnInvalidCondition() throws Exception {
                 .build());
     }
 
-    @Test(expected = NumberFormatException.class)
+    @Test(expected = MessagingException.class)
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/TooManyLinesTest.java]: Or using @Rule ExpectedException?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/RecipientToLowerCaseTest.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Before
+    public void setUp() {
+        testee = new RecipientToLowerCase();
+    }
+
+    @Test
+    public void serviceShouldPutRecipientToLowerCase() throws Exception {
+        FakeMail fakeMail = FakeMail.builder()
+            .recipient(new MailAddress("THienan1234@gmail.com"))
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/RecipientToLowerCaseTest.java]: No it's not (I asked them.)

Putting something close to their email is a way of "signing" their work

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/RecipientToLowerCaseTest.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            .build();
+
+        testee.service(fakeMail);
+
+        Collection<MailAddress> recipients = fakeMail.getRecipients();
+
+        assertThat(recipients)
+            .extracting(AS_STRING)
+            .containsOnly("thienan1234@gmail.com");
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/RecipientToLowerCaseTest.java]: Nope

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java (hunk 8 above)
Comment (chibenwa) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/user/InMemorySubscriptionMapperTest.java]: You are commenting an outdated diff

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+		}
+
+		if(mail.getMessage().getLineCount() > maximumLineCount){
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooMuchLines.java]: expectedException.expect(MessagingException.class);

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooManyLines.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+import javax.mail.MessagingException;
+
+import org.apache.mailet.Mail;
+import org.apache.mailet.MailAddress;
+import org.apache.mailet.base.GenericMatcher;
+
+import com.google.common.collect.ImmutableList;
+
+public class TooManyLines extends GenericMatcher {
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooManyLines.java]: Indent seems strange on this file, no?

Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooManyLines.java]: Seems tab based :)

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/RecipientToLowerCase.java [hunk 15]
```diff
@@ -38,12 +37,13 @@
 
     @Override
     public void service(Mail mail) throws MessagingException {
-        Iterator<MailAddress> rcpts = mail.getRecipients().iterator();
-        List<MailAddress> newRcpts = new ArrayList<MailAddress>();
-        while(rcpts.hasNext()) {
-            newRcpts.add(new MailAddress(rcpts.next().toString().toLowerCase(Locale.US)));
+        ImmutableList.Builder<MailAddress> result = ImmutableList.builder();
+
+        for(MailAddress mailAddress : mail.getRecipients()) {
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/RecipientToLowerCase.java]: Outdated diff comment

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/TooManyLines.java (hunk 14 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/TooManyLines.java]: Fixed. Seems like I missed it.

**Comment by chibenwa** [comment]
Just merged

**Pull request closed without merging**